src/python/pants/backend/python/util_rules/interpreter_constraints.py

```diff
@@ -233,8 +233,79 @@ def requires_python38_or_newer(self) -> bool:
             allowed_versions=["3.8", "3.9", "3.10"], prior_version="3.7"
         )
 
+    def partition_by_major_minor_versions(self) -> tuple[InterpreterConstraints, ...]:
+        """Create a distinct InterpreterConstraints value for each CPython major-minor version that
+        is compatible with the original constraints."""
+        if any(req.project_name != "CPython" for req in self):
+            raise AssertionError(
+                "This function only works with CPython interpreter constraints for now."
+            )
+
+        def all_valid_patch_versions(major_minor: str, last_patch: int) -> list[int]:
+            return [
+                p
+                for p in range(0, last_patch + 1)
+                for req in self
+                if req.specifier.contains(f"{major_minor}.{p}")  # type: ignore[attr-defined]
+            ]
+
+        result = []
+
+        def maybe_add_version(major_minor: str, next_major_minor: str, last_patch: int) -> None:
+            valid_patch_versions = all_valid_patch_versions(major_minor, last_patch)
+            if not valid_patch_versions:
+                return
+
+            if len(valid_patch_versions) == 1:
+                result.append(
+                    InterpreterConstraints([f"=={major_minor}.{valid_patch_versions[0]}"])
+                )
+                return
+
+            skipped_patch_versions = _not_in_contiguous_range(valid_patch_versions)
+            first_patch_supported = valid_patch_versions[0] == 0
+            last_patch_supported = valid_patch_versions[-1] == last_patch
+            if not skipped_patch_versions and first_patch_supported and last_patch_supported:
+                constraint = f"=={major_minor}.*"
+            else:
+                min_constraint = (
+                    f">={major_minor}"
+                    if first_patch_supported
+                    else f">={major_minor}.{valid_patch_versions[0]}"
+                )
+                max_constraint = (
+                    f"<{next_major_minor}"
+                    if last_patch_supported
+                    else f"<={major_minor}.{valid_patch_versions[-1]}"
+                )
+                if skipped_patch_versions:
+                    skipped_constraints = ",".join(
+                        f"!={major_minor}.{p}" for p in skipped_patch_versions
+                    )
+                    constraint = f"{min_constraint},{max_constraint},{skipped_constraints}"
+                else:
+                    constraint = f"{min_constraint},{max_constraint}"
+
+            result.append(InterpreterConstraints([constraint]))
+
+        maybe_add_version("2.7", "2.8", 18)
+        maybe_add_version("3.5", "3.6", 11)
+        maybe_add_version("3.6", "3.7", 15)
+        maybe_add_version("3.7", "3.8", 20)
+        maybe_add_version("3.8", "3.9", 20)
+        maybe_add_version("3.9", "3.10", 20)
+        maybe_add_version("3.10", "3.11", 20)
+        maybe_add_version("3.11", "3.12", 20)
```

Comment:
The effect of including these now is that folks using ranges like `>=3.8,<4` will have builds fail looking for copies of `3.11` and `3.12`... is that intentional to guide them to make their ranges narrower, or should we wait until these versions have had final releases before including them?

Reply:
I'm thinking in the followup that we will have an option to warn or error. So, you can either tighten your constraint to not be <4, or switch it to be a warning.